### PR TITLE
TimeseriesPanel: Fix variables in data links

### LINKED ISSUE #56063: Data link not retrieving the right values on time series panel
**What happened**:
I have a dashboard with three variables [_meter, readtime and kwh_total_], one time series panel and one data panel. Both panels use the exact same query. In both panels, in the overrides section, I added field _kwh_total_ and the following link:
`http://x.x.x.x:3000/d/xNTeB744z/data-manipulation?var-meter=${__data.fields.meter}&var-readtime=${__data.fields.readtime}&var-kwh_total=${__data.fields.kwh_total}
`
In both panels, the link points to the same dashboard. The data link in the table is behaving properly. As I move the mouse from one line to the other, the link variables update according to the line that I am on. The time series panel behaves in an inconsistent manner; as I move the mouse to the data point and hover on top of the link, the link variables do update, but most of the time to the wrong values. As you watch the video look at the status line. You will see clearly what I mean.

**What you expected to happen**:
I expecteds the data link in the time series panel to behave in the same manner as in the table panel.

**How to reproduce it (as minimally and precisely as possible)**:
Please watch the video.

**Anything else we need to know?**:
No.

**Environment**:
- Grafana version: **Gafana v9.1.6 (92461d8d1e)**
- Data source type & version: 
      **Amazon Athena - version: 2.1.0**
- OS Grafana is installed on: 
**Linux ip-10-0-1-42.ec2.internal 5.10.135-122.509.amzn2.x86_64 #1 SMP Thu Aug 11 22:41:16 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux**
- User OS & Browser: MacOS / Chrome
- Grafana plugins: Attached file: plugins.json
- Others: None

https://user-images.githubusercontent.com/3796667/193154272-a2497d2b-e486-49b4-afaf-fd8bc9d2d248.mp4

### PR DESCRIPTION
**Pull request by mdvictor**
**What this PR does / why we need it**:
Variables in data links show wrong values in two situations:

* First situation is when the values in fields are stored in an `ArrayVector` that is aligned by the timeseries panel. The pre-aligned vector vs the aligned vector are of different lengths and thus will return erroneous values in the datalinks since the `getLinks()` method wants a field value index to return a value.

* Second situations is when the fields have their values sorted in a `SortedVector`. Here, we need to send `getLinks()` the order index and not the buffered value index so that it will return the correct value in data link variables.

**Which issue(s) this PR fixes**:

Fixes #56063

**Special notes for your reviewer**:
The biggest problem is that the data is processed after being queried in `PanelQueryRunner` (and has field overrides applied along with data links, on the un-aligned frame) and then the frame is processed again in the timeseries panel for alignment. When building a datalink for a datapoint we use `getLinks` which is generated by `getLinksSupplier` in the first processing step, before frame alignment in the panel, thus getting bad data. My solution was to add a way to override the frame in `getLinks` in special cases like this one where the panel changes the frames, but I'm open for ideas.

### CODE REVIEW

**Review by dprokop — COMMENTED** [review]
@mdvictor This is an interesting approach! I wonder, what's the benefit of it compared to my proposal [here](https://github.com/grafana/support-escalations/issues/3702#issuecomment-1254690123)? Wonder if there is something that I cannot see :) I like the proposed solution, but this line `field.getLinks(valueLinkConfig)` feels mind-bending, as suddenly an already contextualized `getLinks` function, is changing its context.

**Comment by mdvictor** [comment]
Hey @dprokop. You are right, I had the same feeling about the context situation with this override. Now, the more I think about your solution the more I think it's the viable way to go though. What stopped me the first time from going that way was the fact that we would need to get data from the `PanelModel` (the `fieldConfig` - https://github.com/grafana/grafana/blob/main/public/app/features/query/state/PanelQueryRunner.ts#L146) into GraphNG utils (https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/GraphNG/utils.ts#L44-L45). 

But I guess that would be pretty simple to do through the `PanelContext`, right?

**Comment by mdvictor** [comment]
Hey @dprokop. I modified the code and think this is the better way, as you suggested. This will also fix the issue for any `Timeseries` panel, so `Candlestick` also. I looked a bit into `BarChart`, since it's also timeseries, but things are different there and the alignment doesn't change the length of the value buffer. 

Tell me what you think about it when you have time!

File: packages/grafana-ui/src/components/GraphNG/GraphNG.tsx [hunk 1]
```diff
@@ -56,6 +58,7 @@ export interface GraphNGProps extends Themeable2 {
   propsToDiff?: Array<string | PropDiffFn>;
   preparePlotFrame?: (frames: DataFrame[], dimFields: XYFieldMatchers) => DataFrame;
   renderLegend: (config: UPlotConfigBuilder) => React.ReactElement | null;
+  replaceVariables?: InterpolateFunction;
```
Comment (zoltanbedi) [inline packages/grafana-ui/src/components/GraphNG/GraphNG.tsx:61]: My only concern is with this prop. I think it shouldn’t be at this low level component but instead in timeseries component. My thinking is that the variable interpolation is a grafana app specific feature that shouldn’t tied to this component. have you tried adding this logic to the timeseries component instead?

**Review by zoltanbedi — COMMENTED** [review]
Other than my concern about the code the fix worked as expected.

File: packages/grafana-ui/src/components/GraphNG/GraphNG.tsx (hunk 1 above)
Comment (mdvictor) [reply packages/grafana-ui/src/components/GraphNG/GraphNG.tsx:61]: I thought the same initially, but then I saw that Candlestick was also affected by this issue and thought to make the fix for all existing (and future) panels that use `Timeseries` component. I understand the concern though and agree with it. I'll make the changes and push the modifations.

**Review by zoltanbedi — APPROVED** [review]
LGTM

**Comment by peawet08** [comment]
Hi,
I had an issue whilst running 9.2.0 with variables.  We used the variables to get values from a MSSQL server and sometimes the variables would be nested within other variable.

The problem manifested in old data being displayed from a stored procedure was actually returned when running the stored procedure direct in the database.  It was as if the variable had cached the data but there is no caching in the free, on-premise version.

The variables were all set to refresh on load as the pages are auto refreshed external to dashboard and no user interacts with them.  If I changed the variables to refresh on time range change, they would update correctly.

I was forced to change all the dashboards to remove variables where possible.  Spent a week changing them around.  

I have today updated to v9.2.3 and the problem has gone away.  the only change that looks almost related is this one.  So is it possible this fix has fixed my issue?

Thanks everyone for your efforts!!

**Comment by mdvictor** [comment]
Hey @peawet08,

This fix was mainly for variables used on the Timeseries Panel within Data Links, so I'm not sure this might be what fixed the issue. 

Nonetheless, I am glad your issue was fixed and we thank you for your feedback and support!

**Pull request merged**